embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/scheduling/SchedulingServiceImpl.kt

```diff
@@ -1,14 +1,162 @@
+@file:Suppress("FunctionOnlyReturningConstant")
+
 package io.embrace.android.embracesdk.internal.delivery.scheduling
 
+import io.embrace.android.embracesdk.internal.clock.Clock
+import io.embrace.android.embracesdk.internal.comms.api.ApiResponse
+import io.embrace.android.embracesdk.internal.comms.api.Endpoint
 import io.embrace.android.embracesdk.internal.delivery.execution.RequestExecutionService
+import io.embrace.android.embracesdk.internal.storage.PayloadReference
+import io.embrace.android.embracesdk.internal.storage.StorageService2
+import io.embrace.android.embracesdk.internal.worker.BackgroundWorker
+import java.io.InputStream
+import java.util.Collections
+import java.util.LinkedList
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.Future
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicBoolean
 
 internal class SchedulingServiceImpl(
-    @Suppress("unused") private val requestExecutionService: RequestExecutionService
+    private val storageService: StorageService2,
+    private val executionService: RequestExecutionService,
+    private val schedulingWorker: BackgroundWorker,
+    private val deliveryWorker: BackgroundWorker,
+    private val clock: Clock
 ) : SchedulingService {
 
+    private val blockedEndpoints: MutableMap<Endpoint, Long> = ConcurrentHashMap()
+    private val sendLoopActive = AtomicBoolean(false)
+    private val activeSends: MutableSet<PayloadReference> = Collections.synchronizedSet(HashSet())
+    private val payloadsToRetry: MutableMap<PayloadReference, RetryInstance> = ConcurrentHashMap()
+
+    override fun onPayloadIntake() {
+        // When a payload arrives, check to see if there's already an active job try to deliver payloads
+        // If not, schedule job. If so, do nothing.
+        if (sendLoopActive.compareAndSet(false, true)) {
```

Comment:
One trade-off of this approach is that if the `storageService` is saturated with less important payloads such as logs, any new important payloads such as crashes/sessions might not get scheduled for a long time. In extreme cases where the cache limit is reached, that will lead to data loss of important payloads rather than less important ones.

IMO the `SchedulingService` should probably recheck what files are available whenever `onPayloadIntake` is called to avoid this scenario, given that the aim of this project is to avoid all avoidable sources of data loss. This doesn't necessarily have to mean disk writes each time if the `StorageService` maintains an internal list that is updated whenever a file is added/deleted.

Reply:
Hmm. Yeah, that's good to support. I think I'll do something a bit different - every call where there's an active loop, I'll notify it to requery for additional files, basically resetting the loop. This might even allow me to remove the additional check after the loop finishes to pick up newly inserted payloads.

Let me do that!